onyo/cli/cat.py

```diff
@@ -47,11 +54,17 @@ def cat(args: argparse.Namespace) -> None:
     r"""
     Print the contents of **ASSET**\ s to the terminal.
 
-    If any of the paths are invalid, then no contents are printed and an error
-    is returned.
+    If any of the paths are not assets, no content is printed and an error is
+    returned.
+
+    Assets with invalid content are printed with an error message. See **Exit
+    Status** for more information about return codes.
     """
     paths = [Path(p).resolve() for p in args.asset]
 
     inventory = Inventory(repo=OnyoRepo(Path.cwd(), find_root=True))
-    onyo_cat(inventory,
-             paths)
+    try:
+        onyo_cat(inventory,
+                 paths)
+    except InvalidAssetError:
+        sys.exit(1)
```

Comment:
This seems inconsistent with the rest of onyo. `InvalidAssetError` should bubble up and be caught in main. Or is there a reason to not go through that place that would usually decide how to render a message etc.?

Reply:
This is actually intentional. It is a special case, just like `onyo get`. All other errors will exit `2` in `main.py`. This unique case for `cat` is caught, and handled by exiting `1`.

Reply:
Could you maybe add a note as a comment that it is intentional? :)
Things like that will definitely be forgotten by me when I look at the code in some months.

Reply:
If Ben doesn't have a suggestion for a better way of approaching this, I am happy to add a comment stating its intentionality. :-)

Reply:
See https://github.com/psyinfra/onyo/pull/702/files#r1818553447

Reply:
I would, of course, suggest the same thing for `get`. Simply missed that as I didn't get to review the get PR before it was merged.

Reply:
I have fixed this, and it was much more of a rabbit hole than I expected. ;-)

I have updated the PR with three additional commits:
- `ui.error` no longer requires using the message of the exception
- add `OnyoCLIExitCode` exception
- switch `onyo get` over to `OnyoCLIExitCode`

Feedback much appreciated.

Reply:
I have pushed again, this time splitting the traceback-to-string functionality into its own function. There already was a `ui.log_debug()`, that just wasn't being used by `ui.error()`.

I'm not thrilled with the name `format_traceback()`, but felt my other ideas weren't any better (`get_traceback_as_string()`, `traceback_to_string()`, etc). Ideas welcome here.

Reply:
The `pyre `failures are new and unrelated to this PR.

I will investigate them but currently cannot reproduce locally.